examples/DataApi-Voting-poll/template/modules/shared_data/src/main/scala/com/my/currency/shared_data/Main.scala

```diff
@@ -0,0 +1,360 @@
+      case Some(value) =>
+        val pollValidOptions = value.pollOptions
+        pollValidOptions.get(option) match {
+          case Some(_) => ().validNec
+          case None => InvalidOption.asInstanceOf[DataApplicationValidationError].invalidNec
+        }
+      case None => InvalidOption.asInstanceOf[DataApplicationValidationError].invalidNec
+    }
+  }
+
+  def validateProvidedAddress(proofAddresses: List[Address], address: Address): DataApplicationValidationErrorOr[Unit] = {
+    if (proofAddresses.contains(address)) {
+      ().validNec
+    } else {
+      InvalidAddress.asInstanceOf[DataApplicationValidationError].invalidNec
+    }
+  }
+
+  def validateIfUserAlreadyVoted(maybeState: Option[PollState], address: Address): DataApplicationValidationErrorOr[Unit] = {
+    maybeState match {
+      case Some(value) =>
+        val pollVotes = value.votes
+        pollVotes.get(address) match {
+          case Some(_) => RepeatedVote.asInstanceOf[DataApplicationValidationError].invalidNec
+          case None => ().validNec
+        }
+      case None => ().validNec
+    }
+  }
+
+  def validateSnapshotCreatePoll(snapshotOrdinal: SnapshotOrdinal, update: CreatePollUpdate): DataApplicationValidationErrorOr[Unit] = {
+    if (update.endSnapshotOrdinal < snapshotOrdinal.value.value) {
+      InvalidEndSnapshot.asInstanceOf[DataApplicationValidationError].invalidNec
+    } else {
+      ().validNec
+    }
+  }
+
+  def validatePollSnapshotInterval(lastSnapshotOrdinal: SnapshotOrdinal, currentState: State, voteInPoll: VoteInPollUpdate): DataApplicationValidationErrorOr[Unit] = {
+    val poll = currentState.polls.get(voteInPoll.pollId)
+    poll match {
+      case Some(value) =>
+        if (value.endSnapshotOrdinal < lastSnapshotOrdinal.value.value) {
+          ClosedPool.asInstanceOf[DataApplicationValidationError].invalidNec
+        } else {
+          ().validNec
+        }
+      case None => PollDoesNotExists.asInstanceOf[DataApplicationValidationError].invalidNec
+    }
+  }
+}
+
+object Validations {
+  def createPollValidations(update: CreatePollUpdate, state: State, lastSnapshotOrdinal: Option[SnapshotOrdinal]): IO[DataApplicationValidationErrorOr[Unit]] = {
+    val validateCreatePollSnapshot = IO {
+      lastSnapshotOrdinal match {
+        case Some(value) => validateSnapshotCreatePoll(value, update)
+        case None => ().validNec
+      }
+    }
+
+    val validatePoll = IO {
+      val voteId = Hash.fromBytes(customUpdateSerialization(update))
+      validateIfPollAlreadyExists(state.polls.get(voteId.toString))
+    }
+
+    for {
+      validatedCreatePollSnapshot <- validateCreatePollSnapshot
+      validatedPoll <- validatePoll
+    } yield validatedCreatePollSnapshot.productR(validatedPoll)
+
+  }
+
+  def voteInPollValidations(update: VoteInPollUpdate, state: State, lastSnapshotOrdinal: Option[SnapshotOrdinal]): IO[DataApplicationValidationErrorOr[Unit]] = {
+    val validateSnapshotInterval = IO {
```

Comment:
For a lot of other places in the code, we're constructing an IO type here to be consistent with outer functions that require an F type. Since it appears most of the uses here don't actually require an IO in the sense that they're not dealing with a side effect like reading a file (and are instead doing something instantaneous like checking a map or applying a comparison,) it's not really appropriate to use an IO for that. Even if we wanted to compose them, it would be better to use an arbitrary F type, but I wouldn't recommend that either.

Instead, what you should do here is just process the simple logic as is here (without any IO types,) and only at the immediate point that you require it to be used for compatibility with the outer functions, then construct an IO there.

You can also use IO.pure if it's an immediate calculated value (instead of a longer running function which is what the thunk is for.) I.e. if the value has ALREADY been materialized (for instance invoking X > 5 and then assigning to a value,) you can use pure, if the thunk you're passing to IO.apply here (the syntax for (IO{...}) is an actual calculation, then its appropriate to use IO.apply -- but not for already calculated values

Reply:
Thanks for the tip man, updated